lib/control-connection.js

```diff
@@ -75,102 +77,118 @@ function ControlConnection(options, profileManager, context) {
- * @param {Function} callback
+ * Stores the contact point information and what it resolved to.
+ * @param {String|null} address
+ * @param {String} port
+ * @param {String} name
+ * @param {Boolean} isIPv6
  */
-ControlConnection.prototype.init = function (callback) {
-  if (this.initialized) {
-    //prevent multiple serial initializations
-    return callback();
+ControlConnection.prototype.addContactPoint = function(address, port, name, isIPv6) {
+  if (address === null) {
+    // Contact point could not be resolved, store that the resolution came back empty
+    this._resolvedContactPoints.set(name, utils.emptyArray);
+    return;
   }
-  const self = this;
 
-  function addHost(address, port, name, isIPv6, cb) {
-    port = port || self.options.protocolOptions.port;
-    const endPoint = `${address}:${port}`;
-    const h = new Host(endPoint, self.protocolVersion, self.options, self.metadata);
-    self.hosts.set(endPoint, h);
+  const portNumber = parseInt(port, 10) || this.options.protocolOptions.port;
+  const endpoint = `${address}:${portNumber}`;
+  this._contactPoints.add(endpoint);
 
-    let resolvedAddressedByName = self._resolvedContactPoints.get(name);
-    if (resolvedAddressedByName === undefined) {
-      resolvedAddressedByName = [];
-      self._resolvedContactPoints.set(name, resolvedAddressedByName);
-    }
+  // Use RFC 3986 for IPv4 and IPv6
+  const standardEndpoint = !isIPv6 ? endpoint : `[${address}]:${portNumber}`;
 
-    // Use RFC 3986 for IPv4 and IPv6
-    resolvedAddressedByName.push(!isIPv6 ? endPoint : `[${address}]:${port}`);
+  let resolvedAddressedByName = this._resolvedContactPoints.get(name);
+  if (resolvedAddressedByName === undefined) {
+    resolvedAddressedByName = [];
+    this._resolvedContactPoints.set(name, resolvedAddressedByName);
+  }
+
+  resolvedAddressedByName.push(standardEndpoint);
+};
 
-    self.log('info', 'Adding host ' + endPoint);
-    if (cb) {
-      cb();
+ControlConnection.prototype.parseEachContactPoint = function(name, next) {
+  let addressOrName = name;
+  let port = null;
+
+  if (name.indexOf('[') === 0 && name.indexOf(']:') > 1) {
+    // IPv6 host notation [ip]:port (RFC 3986 section 3.2.2)
+    const index = name.lastIndexOf(']:');
+    addressOrName = name.substr(1, index - 1);
+    port = name.substr(index + 2);
+  } else if (name.indexOf(':') > 0) {
+    // IPv4 or host name with port notation
+    const parts = name.split(':');
+    if (parts.length === 2) {
+      addressOrName = parts[0];
+      port = parts[1];
     }
   }
 
-  utils.series([
-    function resolveNames(next) {
-      utils.each(self.options.contactPoints, function eachResolve(name, eachNext) {
-        if (name.indexOf('[') === 0 && name.indexOf(']:') > 1) {
-          // IPv6 host notation [ip]:port (RFC 3986 section 3.2.2)
-          const index = name.lastIndexOf(']:');
-          return addHost(name.substr(1, index - 1), name.substr(index + 2), name, true, eachNext);
```

Comment:
To make sure I understand right:

* Previously, every resolved name for a contact point would create a `Host` entry in `HostMap`.
* With this change, `Host`s will only be created when resolving data from the system.local and peers tables, meaning we will only see `Host` entries for what C* nodes report are nodes.

This makes complete sense to me.  I always thought it was weird that we created hosts for contact points (most other drivers do this as well).

Reply:
yes! that's correct.